File: greedy/java/src/Boats.java
import java.util.Arrays;

class Boats {
	public static void main(String[] args) {
		System.out.println(numRescueBoats(new int[]{1, 2}, 3));
		System.out.println(numRescueBoats(new int[]{3, 2, 2, 1}, 3));
		System.out.println(numRescueBoats(new int[]{3, 5, 3, 4}, 5));
	}

    public static int numRescueBoats(int[] people, int limit) {
        int l = 0, r = people.length - 1;

        Arrays.sort(people);

        int boats = 0;
        while(l <= r) {
            // fit two people in one boat
            if(people[r] + people[l] <= limit){
                l++;
                r--;
            }
            // only fit the heaviest
            else if(people[r] <= limit){
                r--;
            }
            else{
                l++;
            }
            boats++;
        }

        return boats;
    }
}

File: greedy/python/boats.py
class Solution:
    def numRescueBoats(self, people, limit):
        people.sort()
        l, r = 0, len(people) - 1

        boats = 0
        while l <= r:
            # tenta encaixar duas pessoas no mesmo bote
            if people[r] + people[l] <= limit:
                l += 1
                r -= 1
            # caso contrário, coloca apenas a pessoa mais pesada no bote
            elif people[r] <= limit:
                r -= 1
            else:
                l += 1
            boats += 1

        return boats
    
s = Solution()
print(s.numRescueBoats([1, 2], 3))
print(s.numRescueBoats([3, 2, 2, 1], 3))
print(s.numRescueBoats([3, 5, 3, 4], 5))